include/exadg/functions_and_boundary_conditions/evaluate_functions.h

```diff
@@ -217,11 +250,131 @@ struct FunctionEvaluator<1, dim, Number>
         array[v] = function_with_normal->value(q_point, d);
       }
-      value[d].load(&array[0]);
+      value[d].load(array.data());
+    }
+
+    return value;
+  }
+};
+
+template<int dim, typename Number>
+struct FunctionEvaluator<2, dim, Number>
+{
+  static inline DEAL_II_ALWAYS_INLINE //
+    dealii::Tensor<2, dim, dealii::VectorizedArray<Number>>
+    value(std::shared_ptr<dealii::Function<dim>>                      function,
+          dealii::Point<dim, dealii::VectorizedArray<Number>> const & q_points,
+          double const &                                              time)
+  {
+    dealii::Tensor<2, dim, dealii::VectorizedArray<Number>> value;
+
+    for(unsigned int d1 = 0; d1 < dim; ++d1)
+    {
+      for(unsigned int d2 = 0; d2 < dim; ++d2)
+      {
+        std::array<Number, dealii::VectorizedArray<Number>::size()> array;
+        for(unsigned int v = 0; v < dealii::VectorizedArray<Number>::size(); ++v)
+        {
+          dealii::Point<dim> q_point;
+
+          for(unsigned int d = 0; d < dim; ++d)
+            q_point[d] = q_points[d][v];
+
+          function->set_time(time);
+          // Assuming row-major vector function equivalent
+          array[v] = function->value(q_point, d1 * dim + d2);
+        }
+        value[d1][d2].load(array.data());
+      }
+    }
+  }
+
+  static inline DEAL_II_ALWAYS_INLINE //
+      dealii::Tensor<2, dim, dealii::VectorizedArray<Number>>
+      value(std::shared_ptr<FunctionCached<2, dim>> function,
+            unsigned int const                      face,
+            unsigned int const                      q,
+            unsigned int const                      quad_index)
+  {
+    dealii::Tensor<2, dim, dealii::VectorizedArray<Number>> value;
+
+    std::array<dealii::Tensor<2, dim, Number>, dealii::VectorizedArray<Number>::size()>
+      tensor_array;
+
+    for(unsigned int v = 0; v < dealii::VectorizedArray<Number>::size(); ++v)
+      tensor_array[v] = function->tensor_value(face, q, v, quad_index);
+
+    for(unsigned int d1 = 0; d1 < dim; ++d1)
+    {
+      for(unsigned int d2 = 0; d2 < dim; ++d2)
+      {
+        std::array<Number, dealii::VectorizedArray<Number>::size()> array;
+        for(unsigned int v = 0; v < dealii::VectorizedArray<Number>::size(); ++v)
+          array[v] = tensor_array[v][d1][d2];
+
+        value[d1][d2].load(array.data());
+      }
+    }
+  }
+
+  static inline DEAL_II_ALWAYS_INLINE //
+    dealii::SymmetricTensor<2, dim, dealii::VectorizedArray<Number>>
+    value_symmetric(std::shared_ptr<dealii::Function<dim>>                      function,
+                    dealii::Point<dim, dealii::VectorizedArray<Number>> const & q_points,
+                    double const &                                              time)
+  {
+    dealii::SymmetricTensor<2, dim, dealii::VectorizedArray<Number>> value;
+
+    for(unsigned int d1 = 0; d1 < dim; ++d1)
+    {
+      for(unsigned int d2 = d1; d2 < dim; ++d2)
```

Comment:
You are doing here more work than necessary. I would keep it like it is but add some asserts that if you compute an entry twice the value set the second time has to be identical to the first value set.

Reply:
I don't quite get what you mean. The two loops result in going just once over all the entries in the upper triangular, which is in the end as long as the number of independent components, or am I overseeing something? Can you ellaborate?

Reply:
Oh, my mistake. I normally write `for(unsigned int d2 = 0; d2 <= 1; ++d2)`